OpenRA.Mods.Common/Traits/World/WeatherOverlay.cs

```diff
@@ -0,0 +1,327 @@
+							Color = tempColor,
+							TailColor = tempColorTail
+						});
+				}
+			}
+		}
+
+		void ParticlesCountLogic(WorldRenderer wr)
+		{
+			// Logic
+			if (particleCountFader == ParticleCountFaderType.Hold && ParticleList.Count < maxParticleCount)
+				particleCountFader = ParticleCountFaderType.FadeIn;
+			else if (particleCountFader == ParticleCountFaderType.FadeIn && ParticleList.Count >= maxParticleCount)
+				particleCountFader = ParticleCountFaderType.Hold;
+			else if (particleCountFader == ParticleCountFaderType.FadeOut && ParticleList.Count == 0)
+				particleCountFader = ParticleCountFaderType.Hold;
+
+			// Do the fade functions
+			if (particleCountFader == ParticleCountFaderType.FadeIn)
+				FadeInParticleCount(wr);
+			else if (particleCountFader == ParticleCountFaderType.FadeOut)
+				FadeOutParticleCount(wr);
+		}
+
+		void FadeInParticleCount(WorldRenderer wr)
+		{
+			SpawnParticles(1, 0, 100);
+
+			// Remove Particles, which are getting replaced from the top to the bottom by the "EdgeCheckReplace",
+			// when scrolling down, as long as the FadeIn is not completed,
+			// to avoid having particles at the top and bottom, but not in the middle of the screen.
+			for (var i = 0; i < ParticleList.Count; i++)
+			{
+				if (ParticleList[i].PosY < 0)
+					ParticleList.RemoveAt(i);
+			}
+
+			// Add Particles when the weather is fading in and scrolling up, to fill areas above
+			if (antiScrollPrevTopLeft.Y > wr.Viewport.TopLeft.Y)
+			{
+				// Get delta Y and limit to the max value
+				var tempRangeY = antiScrollPrevTopLeft.Y - wr.Viewport.TopLeft.Y;
+				var tempParticleCount = CalculateParticleCount(Game.Renderer.Resolution.Width, (int)tempRangeY);
+				if (ParticleList.Count + tempParticleCount > maxParticleCount)
+					tempParticleCount = maxParticleCount - ParticleList.Count;
+
+				SpawnParticles(tempParticleCount, (int)tempRangeY, 50);
+			}
+		}
+
+		void FadeOutParticleCount(WorldRenderer wr)
+		{
+			for (var i = 0; i < ParticleList.Count; i++)
+			{
+				if (ParticleList[i].PosY > (Game.Renderer.Resolution.Height - ParticleList[i].Gravity))
+					ParticleList.RemoveAt(i);
+			}
+		}
+
+		Particle XAxisSwing(Particle tempParticle)
+		{
+			// Direction turn
+			if (tempParticle.SwingOffset < -tempParticle.SwingAmplitude || tempParticle.SwingOffset > tempParticle.SwingAmplitude)
+				tempParticle.SwingDirection = -tempParticle.SwingDirection;
+
+			// Performe the X-Axis-Swing
+			tempParticle.SwingOffset += tempParticle.SwingDirection * tempParticle.SwingSpeed;
+			return tempParticle;
+		}
+
+		public void Tick(Actor self)
+		{
+			windTickCountdown--;
+		}
+
+		Particle WindLogic(Particle tempParticle)
+		{
+			if (!info.ChangingWind)
+				targetWindXOffset = info.WindStrength[0];
+			else if (windTickCountdown <= 0)
+			{
+				windTickCountdown = Game.CosmeticRandom.Next(info.WindTick[0], info.WindTick[1]);
+				if (Game.CosmeticRandom.Next(1) == 1 && currentWindIndex > 0)
```

Comment:
.Next returns at most 1 less than the given argument, so this will always be false.